### DISCUSSION TITLE
Not installing on Arduino IDE

### DISCUSSION THREAD

**Discussion by joeinaustin** [SpenceKonde/ATTinyCore#838] [Library discussion]
It fails both using the json link in board manager and downloading the zip file.  This is my 1st attempt to use this lib now labeled 2.0 dev.  Any ideas why it fails to install?

**Comment by joeinaustin** [comment]
Eror trying to use zip and the 1.4 version I downloaded ... "Error: 13 INTERNAL: Library install failed: moving extracted archive to destination dir: library not valid"  I also tried the DEV 2.x version, same error.

**Reply by SpenceKonde** [reply]
Per manual installation docs:
```
Download the .zip, extract, and place in the hardware folder inside your sketchbook folder (if there is no hardware folder, create it). You can find/set the location of the sketchbook folder in the Arduino IDE at File > Preferences -> Sketchbook location.
```
Typically on windows will be in a directory like
```
C:\Users\Spence\Documents\Arduino\ (this is the sketchbook)
so you create
C:\Users\Spence\Documents\Arduino\hardware
and copy the extracted folder to there, so you have a 
C:\Users\Spence\Documents\Arduino\hardware\ATTinyCore 
``` 
This directory must contain a folder named avr - if there's an extra layer of folders (eg, `C:\Users\Spence\Documents\Arduino\hardware\ATTinyCore\ATTinyCore-master\avr\...` ), you need to remove extra layer of folders. Some unzipping tools add an extra layer of containing folders whenever unzipping something. Some don't.

A core is not a library. A library is something that you #include the header for, in so doing getting access to the functions declared in that header (typically implemented in an accompanying .cpp file). It does not define the architecture, or provide implementations of the core Arduino API - which a core does. Technically ATTinyCore and everything else like that is not just a core either; the official term is "Third Party Hardware Package", and it includes board definitions, platform definitions, programmer definitions, variant pins_arduino.h to support the board definitions it comes with, as well as the core proper (the part that implements, stuff like digitalWrite and other API stuff), plus a package of platform-specific libraries (Wire.h, SPI.h, EEPROM.h, that kind of thing), some bootloader binaries and bootloader source. 

Anyway, the point is, it's not a library, and telling the IDE to use a zipped library and pointing it to a zipped hardware package won't work - it only knows what to do with a single zipped library, not a whole zipped core. 

As for why installing 1.5.x fails via board manager - Are you using version 2.2.1 of the IDE? That version has a bug they introduced back in august (unrelated to their fixing another bug that I considered a dealbreaker - but they just can't stop fucking with stuff that ain't broke long enough to get a clean release out, so in messing with something that was working fine, now it only works for third party hardware packages that did **not** follow the guidelines for file naming of the json file which *they* specified in the platform specification documentation. Note also that I cannot change the name of the json file (the cause of the problem), because the filename is how they track what packages are associated with what json file, so that would break everything related to my cores for everyone who installed them. As I said this bug was introduced in august and they had a fix in september. They still have not felt a need to do a release, so there's no fixed version available for download. And considering how long it's been, it's a near certainty that the next version will be a mess - there weren't 4 months of bugs to fix, so what could they have been spending the time on if not messing with working functionality so that it will be busted in the next release? 

**No 2.x.x version of the IDE that is fit for use has been released** - the last usable IDE version to my knowledge is 1.8.19, and the one I use is 1.8.13 (though I need to be sure not to compile any malformed code written by untrusted actors, because 1.8.13 has the log4js security vulnerability; malicious content output through that logging facility can be used to execute code with your user privilege level. But by the time they'd fixed that, they had accidentally caused the IDE to not compile any code for any hardware package if any installed hardware package does certain unusual things in platform.txt (I happen to rely on being able to do that during development though). So the most recent viable version of the IDE for me is still the one from 2020 Q2.... 

Arduino just does not seem to be able to keep people from fucking with shit that's working perfectly fine until they break it, and this happens as they're doing bugfixes. Their CI testing is so haphazard that when this broke, I got complaints from two people who test the libraries they maintain against my cores (the compile examples action we all use uses that to install the core) and they thought I'd broken something in my core! But anyway, by the time they've fixed the bugs are are ready for release, they've introduced new ones too. I'm still on 1.8.13, because that's the most recent version not impacted by at least one introduced bug that causes problems for my workflow; for most people, 1.8.19 should work, and things older than 1.8,13 are not handled gracefully. Once it seemed like 1.8.13 got decent uptake, I pulled out the unsightly and distracting hack we were using to make 1.8.12 and earlier behave in a usable way - I'd wanted to drive a stake through it's heart since the day I'd put it in there, in the 1.6.x days.

**Reply by joeinaustin** [reply]
Well SHIT!!  I just updated 3 systems to 2.21 IDE!  I had been working for days on a few projects with no problem util this.  I had no idea Arduino was in such a mess.  I will try the manual method in hopes that things get fixed and you are able to move to 2.x.  I am already getting update errors from the new IDE because I have manually installed multiple copies and version of libs for different projects and it just gives up on updates when you do that.  I have not had great luck relying on older versions of anything so I hope you get it sorted out.  Wish I could help you out somehow.  Thanks for getting back to me.

**Reply by joeinaustin** [reply]
Just one other comment here.  Since I have successfully ben using 221 for a bit now, and added many cores and libs, how did others avoid this issue?   I have only a basic understanding of how they integrate cores, and only beacuse of your description of it (thanks!).  But what you said makes sense.  I tried to add a core as a lib only because I could think of nothing else to do.  I never had a reason to care before.  So from a less informed user perspective, 221 is mostly working for me, and now I wonder if I should backtrack.   I am just thinking out loud here.  Should I be pushing Arduino to fix something in 221?  Waiting?  Loading up 1.8.13??  Can't help but wonder which would be wasting the least time today and in the future.

**Reply by joeinaustin** [reply]
I should have asked this earlier.  Will the manual method work for 221??  The folders for 221 are different as you probably know.  I did find a hardware folder in a completely different place.  ....."\Documents\ArduinoData\packages\arduino\hardware".  Am I wasting my time?  Just putting it in the hardware folder for 221 did not work.

**Comment by hmeijdam** [comment]
> how did others avoid this issue?

I have 17 different [portable installations](https://docs.arduino.cc/software/ide-v1/tutorials/PortableIDE/) of the Arduino IDE 1.x on my PC. Also the main reason I can't go to V2.x It has no option to do a portable install. All these portable installations are tweaked and tuned to a specific hardware or library configuration.
Serves my needs for years now.
